lib/urlapi.c

```diff
@@ -1307,6 +1307,178 @@ CURLU *curl_url_dup(const CURLU *in)
   return NULL;
 }
 
+static char *normalize_path(const char *path)
+{
+  /* split the path per slash, URL decode + encode, then put together again */
+  size_t len = strlen(path);
+  char *sl;
+  struct dynbuf dupe;
+  Curl_dyn_init(&dupe, len * 3);
+
+  do {
+    char *opath;
+    char *npath;
+    int olen;
+    size_t partlen;
+    sl = memchr(path, '/', len);
+    partlen = sl ? (size_t)(sl - path) : len;
+
+    if(partlen) {
+      /* First URL decode the part */
+      opath = curl_easy_unescape(NULL, path, (int)partlen, &olen);
+      if(!opath) {
+        Curl_dyn_free(&dupe);
+        return NULL;
+      }
+
+      /* Then URL encode it again */
+      npath = curl_easy_escape(NULL, opath, olen);
+      curl_free(opath);
+      if(!npath) {
+        Curl_dyn_free(&dupe);
+        return NULL;
+      }
+
+      if(Curl_dyn_add(&dupe, npath) || (sl && Curl_dyn_addn(&dupe, "/", 1))) {
+        curl_free(npath);
+        return NULL;
+      }
+      curl_free(npath);
+    }
+    else if(sl) {
+      /* zero length part but a slash */
+      if(Curl_dyn_addn(&dupe, "/", 1))
+        return NULL;
+    }
+    else {
+      /* no part, no slash */
+      break;
+    }
+
+    if(sl) {
+      path = sl + 1;
+      len -= partlen + 1;
+    }
+
+  } while(sl);
+
+  return Curl_dyn_ptr(&dupe);
+}
+
+/* leave these as-is in query parts */
+#define ISQUERYOK(x) (((x) == ';') || ((x) == '&') || ((x) == '=') ||   \
+                      ((x) == ':') || ((x) == '/') || ((x) == '?') ||   \
+                      ((x) == '+'))
+
+static CURLUcode normalize_query(const char *str, char **updated)
+{
+  /* to handle ' ' to '+' escaping we cannot use libcurl's URL encode
+     function */
+  struct dynbuf dupe;
+  size_t qlen = strlen(str);
+  CURLcode result = CURLE_OK;
+  if(!qlen) {
+    *updated = strdup("");
+    return *updated ? CURLUE_OK : CURLUE_OUT_OF_MEMORY;
+  }
+  Curl_dyn_init(&dupe, qlen * 3);
+
+  while(!result && qlen--) {
+    /* decode then encode */
+
+    /* treat the characters unsigned */
+    unsigned char in = (unsigned char)*str++;
+
+    if(('%' == in) && (qlen >= 2) && ISXDIGIT(str[0]) && ISXDIGIT(str[1])) {
+      curl_off_t byte;
+      char buf[4] = {'%'};
+      const char *p = &buf[1];
+      buf[1] = Curl_raw_toupper(str[0]);
+      buf[2] = Curl_raw_toupper(str[1]);
+      buf[3] = 0;
+      (void)Curl_str_hex(&p, &byte, 255);
+      if(!ISQUERYOK(byte) && (byte > 0x20) && (byte < 0x7f)) {
+        unsigned char num = (unsigned char)byte;
+        result = Curl_dyn_addn(&dupe, &num, 1);
+      }
+      else
+        /* add the original hex in uppercase */
+        result = Curl_dyn_addn(&dupe, buf, 3);
+      str += 2;
+      qlen -= 2;
+      continue;
+    }
+
+    if(in == ' ')
+      result = Curl_dyn_addn(&dupe, "+", 1);
+    else if(ISALNUM(in) || ISURLPUNTCS(in) || ISQUERYOK(in))
```

Comment:
```suggestion
    else if(ISUNRESERVED(in) || ISQUERYOK(in))
```

This seems more obvious to me.